src/optimizers/quantizer.cpp

```diff
@@ -0,0 +1,187 @@
+#include <cmath>
+
+#include "optimizers/quantizer.h"
+#include "tensors/tensor_allocator.h"
+#include "tensors/tensor_operators.h"
+
+#include "functional/functional.h"
+
+namespace marian {
+
+/* simulate a fixed quantization for values in data.
+ * For example:
+ * data  = [0.96, 0.73, 0.82, 0.84, 0.42, 0.29, 0.65]
+ * res   = [1   , 0.6,  0.8 , 0.8 , 0.4,  0.2 , 0.6 ]
+ *
+ * @param data contains the original data
+ * @param res will contain the resulting quantized data. set data = quant for in-place operation
+ * @param numCenters the number of quantized centers in absolute. It should be 2^(bit-1)
+ * @param S stores the scaling factor.
+ */
+void quantizeFixed(Tensor data, Tensor res, int numCenters, float S) {
+  using namespace functional;
+  float multiplier = numCenters / S;
+
+  // clip based on the scale
+  Element(_1 = clip(_2, S), res, data);
+
+  // get the quantization center
+  Element(_1 = round(_1 * multiplier), res);
```

Comment:
I would read "get the quantization center" as "determine a mean to center quantization around." You are not actually getting the quantization center, but rather the id of the quantization bin. I think you can just say "quantize" here.

Reply:
updated